core/src/main/java/org/ehcache/core/spi/time/TickingTimeSource.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ehcache.core.spi.time;
+
+import java.util.Timer;
+import java.util.TimerTask;
+
+/**
+ * The default {@link TimeSource} in Ehcache. it increases the time every millisecond
+ * using a timer. We set it to the system time every second to prevent time drifting.
+ */
+public class TickingTimeSource implements TimeSource {
+
+  private static final long GRANULARITY_IN_MS = 1L;
+
+  private static final long PERIOD_BETWEEN_SYSTEM_SET_IN_MS = 10_000L;
+
+  /**
+   * Singleton instance of this {@link TimeSource}
+   */
+  public static final TimeSource INSTANCE = new TickingTimeSource();
+
+  private volatile long currentTime;
+  private volatile long lastUpdate;
+
+  private TickingTimeSource() {
+    updateToSystemTime();
+    Timer timer = new Timer("TickingTimeSource-timer", true);
+    timer.scheduleAtFixedRate(new TimerTask() {
+      @Override
+      public void run() {
+        if (currentTime - lastUpdate >= PERIOD_BETWEEN_SYSTEM_SET_IN_MS) {
+          updateToSystemTime();
```

Comment:
This could step back in time... can't we fix that?

Reply:
Really slightly yes. After a discussion with @lorban, we agreed that we don't care.